### DISCUSSION TITLE
serialisation of `std::map<size_t, std::set<size_t>>`

### DISCUSSION THREAD

**Discussion by mobergmann** [nlohmann/json#4070] [Q&A] [answered]
I want to serialize a `std::map<size_t, std::set<size_t>>` and I have the following serialization code: 

```c++
std::map<dpp::snowflake, std::set<dpp::snowflake>> newsletters;

// convert newsletter to json
json j(newsletters);
    
// write json to file
std::ofstream o(newsletters_file_path);
o << std::setw(4) << j << std::endl;
```

Saving the file works, but the resulting json file looks like this: 
```json
[
    [
        "213717396026949632",
        [
            "846874459842936832",
            "847515213158285342"
        ]
    ]
]
```
But I have expected sth like this:
```json
[
    [
        213717396026949632,
        [
            846874459842936832,
            847515213158285342
        ]
    ]
]
```

Is there an (easy) way to serialize the types as size_t and not string

**Comment by nlohmann** [comment]
This is not possible. JSON only allows object keys to be strings.

**Reply by gregmarr** [reply, marked as answer]
> This is not possible. JSON only allows object keys to be strings.

Since this isn't creating JSON objects, but JSON arrays, there should be no problem converting the map to the expected form.  The library itself probably couldn't do it easily due to backwards compatibility issues, but it is possible.

I think the appropriate `from_json` already exists:  https://github.com/nlohmann/json/blob/5d2754306d67d1e654a1a34e1d2e74439a9d53b3/include/nlohmann/detail/conversions/from_json.hpp#L422

I'm guessing that the program would need to handle the entire conversion process itself.  https://github.com/nlohmann/json#how-do-i-convert-third-party-types
